src/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -1269,6 +1340,25 @@ impl<'tcx> GotocCtx<'tcx> {
             _ => unreachable!("Expected a single-variant ADT. Found {:?}", struct_type),
         }
     }
+
+    /// Extract a trait type from a `Struct<dyn T>`.
+    /// Note that `T` must be the last element of the struct.
+    /// This also handles nested cases: `Struct<Struct<dyn T>>` returns `dyn T`
+    pub fn extract_trait_type(&self, struct_type: Ty<'tcx>) -> Option<Ty<'tcx>> {
+        let mut typ = struct_type;
+        while let ty::Adt(adt_def, adt_substs) = typ.kind() {
+            assert_eq!(
+                adt_def.variants().len(),
+                1,
+                "Expected a single-variant ADT. Found {:?}",
+                typ
+            );
+            let fields = &adt_def.variants().get(VariantIdx::from_u32(0)).unwrap().fields;
+            let last_field = fields.last().expect("Trait should be the last element.");
```

Comment:
We'll print warnings for all u supported constructs in every crate we are compiling.

If you want I can try to aggregate the data into one big warning at the end. But it will be a follow up PR

Reply:
Comment in the wrong place. :)

Reply:
This function assumes that this is trait or struct with a trait element. I will add a check to the beginning of the function.